BuildSourceImage.sh

```diff
@@ -1,80 +1,968 @@
+    unset IFS
+
+    for pfx in ${pfxs[@]} ; do
+        if echo ${ref} | grep -q "^${pfx}:" ; then
+            # break when we match
+            echo ${ref}
+            return 0
+        fi
+    done
+    # else default
+    echo "docker://${ref}"
+}
+
+#
+# an inline namer for the source image
+# Initially this is a tagging convention (which if we try estesp/manifest-tool
+# can be directly mapped into a manifest-list/image-index).
+#
+ref_src_img_tag() {
+    local ref="${1}"
+    echo -n "$(parse_img_tag ${ref})${source_image_suffix}"
+}
+
+#
+# call out to registry for the image reference's digest checksum
+#
+fetch_img_digest() {
+    local ref="${1}"
+    ## TODO: check for authfile, creds, and whether it's an insecure registry
+    local dgst=$(skopeo inspect "$(ref_prefix ${ref})" | jq .Digest | tr -d \")
+    local ret=$?
+    if [ $ret -ne 0 ] ; then
+        echo "ERROR: check the image reference: ${ref}" >&2
+        return $ret
+    fi
+
+    echo -n "${dgst}"
+}
+
+#
+# pull down the image to an OCI layout
+# arguments: image ref
+# returns: path:tag to the OCI layout
+#
+# any commands should only output to stderr, so that the caller can receive the
+# path reference to the OCI layout.
+#
+fetch_img() {
+    local ref="${1}"
+    local dst="${2}"
+
+    _mkdir_p "${dst}"
+
+    local base="$(parse_img_base ${ref})"
+    local tag="$(parse_img_tag ${ref})"
+    local dgst="$(parse_img_digest ${ref})"
+    local from=""
+    # skopeo currently only support _either_ tag _or_ digest, so we'll be specific.
+    if [ -n "${dgst}" ] ; then
+        from="$(ref_prefix ${base})@${dgst}"
+    else
+        from="$(ref_prefix ${base}):${tag}"
+    fi
+
+    ## TODO: check for authfile, creds, and whether it's an insecure registry
+    ## destination name must have the image tag included (umoci expects it)
+    skopeo \
+        copy \
+        "${from}" \
+        "oci:${dst}:${tag}" >&2
+    echo -n "${dst}:${tag}"
+}
+
+#
+# upack_img <oci layout path> <unpack path>
+#
+unpack_img() {
+    local image_dir="${1}"
+    local unpack_dir="${2}"
+
+    if [ -d "${unpack_dir}" ] ; then
+        _rm_rf "${unpack_dir}"
+    fi
+
+    # TODO perhaps if uid == 0 and podman is present then we can try it?
+    if [ -z "$(command -v umoci)" ] ; then
+        # can be done as non-root (even in a non-root container)
+        unpack_img_umoci "${image_dir}" "${unpack_dir}"
+    else
+        # can be done as non-root (even in a non-root container)
+        unpack_img_bash "${image_dir}" "${unpack_dir}"
+    fi
+}
+
+#
+# unpack an image layout using only jq and bash
+#
+unpack_img_bash() {
+    local image_dir="${1}"
+    local unpack_dir="${2}"
+
+    # for compat with umoci (which wants the image tag as well)
+    if echo "${image_dir}" | grep -q ":" ; then
+        image_dir="${image_dir%:*}"
+    fi
+
+    local mnfst_dgst="$(cat "${image_dir}"/index.json | jq '.manifests[0].digest' | tr -d \" )"
```

Comment:
```
In BuildSourceImage.sh line 290:
    local mnfst_dgst="$(cat "${image_dir}"/index.json | jq '.manifests[0].digest' | tr -d \" )"
          ^--------^ SC2155: Declare and assign separately to avoid masking return values.
                            ^-----------------------^ SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.
```

Reply:
Same in line 294.